### PR TITLE
Automatic broadcast in basic_math and matmul

### PR DESCRIPTION
**Pull request by beam2d**
This PR makes Variable operators broadcast operands automatically. To make the implementation simple, I added `utils.sum_to` and `F.sum_to`, which computes the gradient of broadcast. Note that other functions that should support automatic broadcast (e.g. `where`) are not covered in this PR. A part of #3968.

### CODE REVIEW

**Comment by toslunar** [comment]
I suggest `expect_broadcast_shapes` should accepts two **or more** shapes, if it'll be used everywhere.
`expect_broadcast_shapes(x1, x2) and expect_broadcast_shapes(x2, x3)` does not imply `(x1, x2, x3)` can be broadcasted, e.g. if their shapes are `(2,)`, `(1,)`, `(3,)`

File: chainer/utils/array.py [hunk 1]
```diff
@@ -35,3 +36,16 @@ def empty_like(x):
         return cuda.cupy.empty_like(x)
     else:
         return numpy.empty_like(x)
+
+
+def sum_to(x, shape):
```
Comment (toslunar) [inline chainer/utils/array.py]: Could you assert `not isinstance(x, Variable)`?  Once `Variable.sum` is implemented, `chainer.utils.sum_to` can compute `Variable`'s "sum_to" but its backward is slower than `F.sum_to`'s.

**Comment by toslunar** [comment]
Could you support broadcast in `F.matmul` when one of the two operands is a vector? 
 See e.g. https://github.com/cupy/cupy/blob/2b56cea0007d2efd0aad2fecb4f543c08d711f9b/tests/cupy_tests/math_tests/test_matmul.py#L22-L23

File: chainer/utils/array.py (hunk 1 above)
Comment (beam2d) [reply chainer/utils/array.py]: Fixed to check if x is not Variable.

**Comment by beam2d** [comment]
- Rewrote `expect_broadcast_shapes` to support more than two inputs. I now find reusing `expect` inside of it is useless, so I just make it directly check the shapes and raise `InvalidType` with a custom error message.
- Make `matmul` support broadcast of a vector operand.

File: chainer/utils/type_check.py [hunk 2]
```diff
@@ -583,3 +585,33 @@ def prod(xs):
...
     else:
         return _prod(xs)
+
+
+def expect_broadcast_shapes(*shape_types, **kwargs):
+    """Checks if two shapes are broadcastable at first some axes.
+
+    Args:
+        shapes_types: Type-checked shapes of the arrays to broadcast.
+        ignore_tail (int): Number of axes at the tail to ignore.
```
Comment (toslunar) [inline chainer/utils/type_check.py]: Should the function check the shapes have tail dimensions?  I prefer `type_check.expect_broadcast_shapes(a_type.shape[:-2], b_type.shape[:-2])` as the meaning of `ignore_tail` is not clear.

File: chainer/functions/math/matmul.py [hunk 3]
```diff
@@ -88,48 +89,85 @@ def check_type_forward(self, in_types):
...
         if a_ndim == 0 or b_ndim == 0:
             pass
-        elif a_ndim == 1 or b_ndim == 1:
-            type_check.expect(
-                a_type.ndim == b_type.ndim,
-                a_type.shape == b_type.shape,
-            )
+        elif a_ndim == 1:
+            if b_ndim == 1:
+                type_check.expect(a_type.shape == b_type.shape)
```
[the remark this answers is not in the document]
Comment (beam2d) [reply chainer/functions/math/matmul.py]: I don't think this case can be eliminated. It's a check for the case `a_ndim == 1 == b_ndim`.

File: chainer/utils/type_check.py (hunk 2 above)
Comment (beam2d) [reply chainer/utils/type_check.py]: Looks better. Fixed accordingly.

**Comment by beam2d** [comment]
@toslunar Sorry for being late. I made a reply and a fix for a comment.

File: chainer/functions/math/matmul.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (toslunar) [reply chainer/functions/math/matmul.py]: I meant `a_ndim` doesn't matter if `b_ndim == 1`:
```
        if a_ndim == 0 or b_ndim == 0:
            pass
        elif b_ndim == 1:
            type_check.expect(a_type.shape[-1] == b_type.shape[0])
        elif a_ndim == 1:
            type_check.expect(a_type.shape[0] == b_type.shape[-2])
        else:
```

[the remark this answers is not in the document]
Comment (beam2d) [reply chainer/functions/math/matmul.py]: Ah, I see. OK, I wil fix it.

**Comment by beam2d** [comment]
Simplified the logic based on the suggestions.

**Comment by beam2d** [comment]
I found that `rollaxis` has a bug in backward. I will fix it in another PR.

**Comment by beam2d** [comment]
Made a fix: #4836.

File: chainer/utils/type_check.py [hunk 4]
```diff
@@ -5,9 +5,11 @@
 import threading
 
 import numpy
+import six
 
 import chainer
 from chainer.backends import cuda
+from chainer.utils import argument
```
Comment (toslunar) [inline chainer/utils/type_check.py]: Remove unused import.

File: chainer/functions/math/matmul.py [hunk 5]
```diff
@@ -88,48 +89,82 @@ def check_type_forward(self, in_types):
...
+            # may broadcast
             y = _matmul(a, b, self.transa, self.transb, self.transc)
-        if self.dtype is not None and y.dtype != self.dtype:
-            y = y.astype(self.dtype)
-        return utils.force_array(y),
+
+        if self.dtype is not None:
+            dtype = self.dtype
+        else:
+            dtype = a.dtype
```
Comment (toslunar) [inline chainer/functions/math/matmul.py]: Is this a typo of `y.dtype`?

File: chainer/utils/type_check.py (hunk 4 above)
Comment (beam2d) [reply chainer/utils/type_check.py]: Fixed.

File: chainer/functions/math/matmul.py (hunk 5 above)
Comment (beam2d) [reply chainer/functions/math/matmul.py]: Oops, you're right. Fixed.

**Comment by toslunar** [comment]
The travis failure seems due to numpy/numpy#10343.  Use `xp.matmul` whenever it exists.

**Comment by beam2d** [comment]
Thank you, I fixed it to use matmul if available.

File: chainer/functions/math/basic_math.py [hunk 6]
```diff
@@ -368,35 +392,55 @@ def forward_cpu(self, inputs):
...
+                gx0 = -ggx1 * gy / x1_square
+                ret.append(chainer.functions.sum_to(gx0, x0.shape))
+
         if 1 in indexes:
-            gx1 = -ggx0 * gy / x1_square + \
-                ggx1 * 2 * gy * x0 / (x1_square * x1)
-            ret.append(gx1)
+            gx1 = None if ggx0 is None else -ggx0 * gy / x1_square
+            gx1_1 = (None if ggx1 is None else
+                     ggx1 * 2 * gy * x0 * (x1_square * x1))
```
Comment (toslunar) [inline chainer/functions/math/basic_math.py]: `ggx1 * 2 * gy * x0 * (x1_square * x1)` should be
`ggx1 * 2 * gy * x0 / (x1_square * x1)`

Comment (beam2d) [reply chainer/functions/math/basic_math.py]: Thanks, I fixed it.

**Comment by toslunar** [comment]
Could you resolve circular imports?

**Comment by beam2d** [comment]
Made a fix for the circular imports.

**Comment by toslunar** [comment]
A GPU test is failing.  Fix the test: #4863.

**Comment by beam2d** [comment]
Rebased with the fix #4863.

**Comment by toslunar** [comment]
`chainer-doc` test warns there's no doc for `chainer.functions.sum_to`.

**Comment by beam2d** [comment]
Added the reference entry of `sum_to`.

**Comment by toslunar** [comment]
@beam2d Could you check if my fix is ok?

**Comment by beam2d** [comment]
The fix looks good.

**Pull request merged**